RavenDB-23085 MemoryMap tracking, IO_Uring, Shared Journals

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23085 

This is a big change that can be reviewed as three separated (built upon each other) PR here:

https://github.com/ayende/ravendb/pull/4697
https://github.com/ayende/ravendb/pull/4696
https://github.com/ayende/ravendb/pull/4695

### Additional description

* Track memory mapped & locked memory properly
* Implement write modes, which allows us to customize how we are writing to the disk

Can be controlled using `RAVEN_WRITER_MODE` or will auto select for writing to the _data file_:

** `vectored_file_io` - using `pwritev` system call
** `file_io` - separate `pwrite` call 
** `io_ring` - using `IO_Uring` / `HIORING` 
** `mmap` - (as we previously did, using `memcpy` to writable memory)

* Shared Journals allow us to massively reduce the number of write calls we use when a database has a lot of indexes


* Notice - this moves `Voron` 's schema version to `24`


### Type of change

- [x] Optimization
- [x] New feature

### How risky is the change?

- [x] High

### Backward compatibility

- [x] Ensured. Please explain how has it been implemented?

Schema migration


### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.

ayende committed:
RavenDB-23085 Properly track MemoryMapped as we map using the PAL

ayende committed:
RavenDB-23085 Locked memory tracking was moved to the C PAL code

ayende committed:
RavenDB-23085 Making the PAL build update the Directory.Packages.props file

ayende committed:
RavenDB-23085
 - Expose memory lock callback to managed code
 - Only lock once if we have WRITEABLE_MAP
 - Keep the file name around in the handle for reporting purposes in Windows as well
 - Fixing computation of _GetNearestFileSize
 - Properly error on Windows if we cannot lock the memory

ayende committed:
RavenDB-23085
 - Moved handling of memory locking failure from native code to managed code
 - Avoid double counting of memory locks on WRITEABLE_MAP files

ayende committed:
RavenDB-23085 - Move registration of file mapping to the Pager.State ctor

ayende committed:
RavenDB-23176 Avoid race condition where we may try to truncate the data file if we are concurrently flushing.
Using CreateFileMapping to ensure that we are always just increase the file size

ayende committed:
RavenDB-23085 - Ensure that the test will only account for _our_ changes, not anything else that may happen

ayende committed:
RavenDB-23085 Reversing polarity of check - mlock returns zero on _success_

ayende committed:
RavenDB-23085 - Reworking how we handle writing to the data pager file
We now support write modes (file I/O, mmap, io ring) to handle flushing data into the data pager

- Added locks around pager operations (creation / disposal and writing)
- We now pass the entire set of pages to be written directly to native code for optimal behavior
- Pager's state was split at the native level to global and handle specific state - with ref counting

ayende committed:
RavenDB-23085 - Added support for IO Ring, vectored file IO on linux, as well as mmap & file i/o in both linux & mac

ayende committed:
RavenDB-23085
- Allow to debug PAL using native debugger in Linux
- Fixing invalid flags & handling of IO_Uring setup

ayende committed:
RavenDB-23085 - Ensuring we properly reference the static liburing

ayende committed:
RavenDB-23085
- Only use IO_RING if we don't pass writable_map (so we'll actually write using this method)
- Better error handling if we can't lock memory for the ring on Linux
- Simplified selecting the appropriate write mode

ayende committed:
RavenDB-23085 - Fixing an issue where writing more than 256 pages at once caused us to skip some

ayende committed:
RavenDB-23085 - Better recovery command in the error message

ayende committed:
RavenDB-23085
- Fixed windows truncate file handling
- Code gen improvement for likely branch

ayende committed:
RavenDB-23085
- Fixing IO error not releasing the lock
- Added locking for IO Ring writes in Linux
- Cached the check for kernel version for io ring

ayende committed:
RavenDB-23085 - Changing how we call IO Ring calls in Windows to use dynamic load, to avoid faulting on older versions

ayende committed:
RavenDB-23085 - Allowing to build PAL using MSVC, not just zig cc

ayende committed:
RavenDB-23085
- Started to implement shared jounrals
- We can now submit a journal write from a branch env to a root env, which will record it in its own journal

ayende committed:
RavenDB-23085 - Simplifying how we track the last flushed state for syncing to the headers files

ayende committed:
RavenDB-23085
- Changing the way we write journals in PAL / native code so we can accept multiple entries to the journal at once
- Ensure that initial setup of the IoRing functions in Windows is thread safe
- Adding a way to generate hard links in PAL

ayende committed:
RavenDB-23085
- Better name for the system store when running in memory
- Updating journal writes to the new API
- Better handling of submission of journal entry to be written by root env
- Separated how we detect if transaction requires writing to the journal
- An empty transaction may still want to write to the journal, if there are pending writes from branch env.
- Adding test showcasing merge journals

ayende committed:
RavenDB-23085
- Implement shared journals queue with a proper error handling / waiting
- Will hard-link files during shared journals to reduce the number of overall journals in use

ayende committed:
RavenDB-23085
- Fixing tests to match actual behavior

ayende committed:
RavenDB-23085
- Adding DatabaseId to TransactionHeader and FileHeader
- Updating Voron schema version
- Refactoring how we read the headers and verify them

ayende committed:
RavenDB-23085
- Added DbId to the header.one and header.two files (in a forward compatible manner) so we can use the db id during recovery
- Reading of the headers files is now done & validated in a single location, and greatly simplified
- Can now register in the TransactionHeader which db it belongs to
-
- Finished refactoring how we deal with header accessor without pointers
- Copy db id to the header when compacting storage

ayende committed:
RavenDB-23085 - Removed journal recycling feature - because we want to use shared journals, and those two are in conflict with one another

ayende committed:
RavenDB-23085
- Updated journal recovery code to take into account shared journals

ayende committed:
RavenDB-23085
- Do not try to open a journal writer if you are not the root journal
- Fixing test assertion

ayende committed:
RavenDB-23085 - To handle recovery and writing to the same journal file after restart, we now check if the file is already a hard link, and the wire it the same journal entry in the branch in that case

ayende committed:
RavenDB-23085 - Ensuring that we can upgrade from a legacy database to the new format seamlessly

ayende committed:
RavenDB-23085
- When we read a legacy journal, we'll not keep using it, but create a _new_ file - so all root environment can immediately share with their branches
- Refactored how we deal with recovery
- Adding a bunch of tests to verify various migration from legacy scenarios

ayende committed:
RavenDB-23085 - Handling recovery & initialization of branch environment

ayende committed:
RavenDB-23085
- Fixing jounral typo
- Avoid repeated values meaning the same thing in env record
- Move updating the journal position from JournalFile to the WriteAheadJournal so we can properly track across shared journals
- Transaction headers are now properly accounted for on a per journal file instance per environment, and not globally
- JournalFile now has (global across envs) flag that tell us if we are done writing to it, so both branch and root can easily discard completed journals
- When there is no active transaction and we are flushing, we take the chance to discard all completed journals immediately, in both branch & root env.

ayende committed:
RavenDB-23085
- Properly resetting the current file when we are done with it

ayende committed:
RavenDB-23085
- Wiring the indexes as branches to the root of the database - so we have a single shared journal and reduced writes.

ayende committed:
RavenDB-23085 - Reversing polarity of check

ayende committed:
RavenDB-23085 - Removing Recyclable Journals remnants

ayende committed:
RavenDB-23085
- Properly handle journal sizes that are not multiple of 8KB but 4KB
- Deal with transaction that ends in the last 4KB of the journal

ayende committed:
RavenDB-23085 - Ensuring we properly cleanup on error in PAL

ayende committed:
RavenDB-23085 - Fixing recovery while dealing with corrupted journals

ayende committed:
RavenDB-23085
- Ensuring that we properly mark journals as done when we recover a database
- Removing debug code

ayende committed:
RavenDB-23085
 - Ensuring that we properly scope the shared journals so we avoid trying to use them after the caller exited the scope

ayende committed:
RavenDB-23085
- Changed disposal order of DocumentDatabase so we'll first dispose of the indexes, and only then the transaction merger, so we'll let the indexes flush their changes to the shared journal
- Ensure that we are properly setting the scope for the database transaction merger

ayende committed:
RavenDB-23085 - Move handling of cancellation to the root journal SubmitBranchJournalEntry

ayende committed:
RavenDB-23085
- Separated TxMerger shutdown token from DatabaseShutdown cancellation token - because we want to stop the transaction merger later
- Transaction merger is now aware that it should run transactions if there is a pending journal write
- Cancellation Token for merged commits is now external to the journal

ayende committed:
RavenDB-23085
- Disable recovery & integrity checks when we are in VerifyNoUnexpectedValidTransactionsAfter
- DisableOnRecoveryErrorHandler & DisableOnIntegrityErrorOfAlreadySyncedDataHandler now use a non-null disabling method, so we won't throw if we detect an error while disabled
- Fixed the test corrupting the journal so it would corrupt the journal from that point _forward_

ayende committed:
RavenDB-23085
- Moving test throw hook to just before actual write to file

ayende committed:
RavenDB-23085 - Fixing IncrementalBackup - will now handle reading of the database id in a snapshot scenario properly

ayende committed:
RavenDB-23085
- Changing to using dedicated JournalId per environment that is not exposed at any higher level
- This allows far simpler mechanism with regards to moving databases & indexes, restoring from snapshot backups, etc

ayende committed:
RavenDB-23085
- Moved join of index journal to database journal in root / branch mode to InitializeOptions so will apply to new indexes from the start
- Added a test to check that we are using hard links for indexes from the get go
- When done with shared journal scope - ensure that we replace the cancellation token with an already canceled one
- Properly handle recovering from journal that has transactions only for a branch and not the root
- Removed debug code that no longer makes sense given root/ branch usage

ayende committed:
RavenDB-23085
- Report on debug endpoint whatever an environment is a root or a branch
- Move hard link detection to the StorageEncryptionOptions
- Use a linked cancellation token source to ensure that we are properly canceled when exiting shared journal state.
- Ensuring that the branch will wait until the root releases it, to avoid accessing released memory
- Fixing tests

ayende committed:
RavenDB-23085
- Fixing test reliance on env variable that isn't reuqired.

ayende committed:
RavenDB-23085
- Fixing snapshot restore - we'll force the creation of a _new_ journal file after to restore, to prevent linking the indexes to an old journal file
- Fixing tests

ayende committed:
RavenDB-23085
- Changed how we handle avoiding merging a shared journal that already contains entries from another env that aren't inline (snapshots taken at different times)
- We now detect if the journal file contains the environment at recovery time and if they do, we only allow to use the journal for that environment if a hard link already exists.

ayende committed:
RavenDB-23085
- Added configuration option to disable shared journals
- Disable shared journals for a test that manually tie between transaction merger commands and indexing progress

ayende committed:
RavenDB-23085
- Set the option AvoidSharedJournals globally for the test

ayende committed:
RavenDB-23085 - Fixing compilation after the rebase

ayende committed:
RavenDB-23085 - Ensuring that we setup the avoid shared journal flag at the database level, not the server. The reason is that during full test run, we may get the global server, without this flag.

ayende committed:
RavenDB-23085
- Fixing tests on old versions of Windows in which we cannot delete files that are in use
- Avoid emitting warning messages for disposing already disposed HttpCache in the finalizer, there is no guaranteed order there

ayende committed:
RavenDB-23085
- Fixing MemoryMapped being 0
- Avoid a lambda capture if the file already exists in the FileMapping dictionary

ayende committed:
RavenDB-23085
- Removed unused parameter in EnvironmentStateRecord

ayende committed:
RavenDB-23085
- Ensure that we properly msync() mmap data on write
- Handle truncated writes for pwrites and pwritev
- Ensure that we are always reading from IORing value using interlocked funcs

ayende committed:
RavenDB-23085 - Properly handle partial writes using IO_Uring by completing them

ayende committed:
RavenDB-23085 - Fixing return value check

ayende committed:
RavenDB-23085 - Instead of trying to dynamically optimize the journal compression ratio, make it configurable (defaults to 1)

ayende committed:
RavenDB-23085
- PR comments - better API parameter order for rvn_writer
- Better error handling if failed to delete recyclable journal file during schema upgrade

ayende committed:
RavenDB-23085- PR comments
- added note where we got the liburing.a files
- Restored debug code when failed to flush to file
- Better error message when cannot create IO Ring on Linux because of no mem

ayende committed:
RavenDB-23085 - Changed env variable to explicitly control the writer mode to RAVEN_VORON_WRITER_MODE

ayende committed:
RavenDB-23085 - Moving check of invalid flushed transaction id from sync time to flush time, to detect the actual error more quickly

ayende committed:
RavenDB-23085
- Fixing an issue where a root branch commit that also included enough branch commits to move to a new journal would cause confusion with regards to which journal we should look at
- We now explicitly track the last journal for the environment that we committed into
- Moved validation code earlier in the process to "catch" such issues in the future.

- Ensured that Raven.License.Path env variable is also viable for running tests

ayende committed:
RavenDB-23085 - Fixing logging statement

ayende committed:
RavenDB-23085 - Allow to disable io_uring using env variables properly

ayende committed:
RavenDB-23085
- Root environment will always process its own journal entries first

ayende committed:
RavenDB-23085
- Refactoring integration of shared journals with the transaction merger.
- The transaction merger can now indicate to the root journal that it has more important writes pending, and that it should close the writes and let the _next_ transaction take over
- This is done to make sure that there is a bounded time for the transaction merger to contribute to other environments before it start working on its own stuff.
- Important to ensure that we have low latency under load

ayende committed:
RavenDB-23085
- Reversing polarity of check for throughput vs. latency in shared journals
- Small refactoring to make naming easier
- Fixing configuration messages

ayende committed:
RavenDB-23574 Add configuration to change journal size

ayende committed:
RavenDB-23085 - Avoid race conditions when setting up the shared journal scope

ayende committed:
RavenDB-23085 - Adjust the size of the journal file based on the number of registered environments in root/branch model

ayende committed:
RavenDB-23085 - Avoid checking if the file is linked if this is a new file (and thus cannot be linked)

ayende committed:
RavenDB-23085 - Removed the CurrentJournal / LastCreatedJournal from the header file - we now use the existence of the journal file instead. This allows us to skip an fsync() call whenever we create a new journal.

ayende committed:
RavenDB-23085
- Will now register a LinkJrnl entry to register the linked journals in the shared transaction journal
- Removed dynamic journal file sizing based on the number of participating environments
- Default journal max size set to 4GB

ayende committed:
RavenDB-23085 - As part of recovery of a root journal, attempt to re-wire hard links to the shared journals from the linked journal record.
- Update API naming for PAL to explicitly reflect that we are calling hard link without expecting durability of the name in the case of a system crash.

ayende committed:
RavenDB-23085 - Change the IO ring impl to use a single global instance for the entire process.
- We have a dedicated thread that we can submit work which will use the IO Ring to do the actual IO
- The dedicated thread can be setup to do low priority IO
- This is meant to handle flushes and syncs

ayende committed:
RavenDB-23085 - Now handle fsync via ioring on linux as well

ayende committed:
RavenDB-23085 - Removing tabs

ayende committed:
RavenDB-23085 - Added configuration options to RavenDB to control write mode and io ring size directly
- Better comments in code to explain intent
- Ensured that we fsync the parent directory of a journal file when we create it
- Fixing partial writes using io ring

ayende committed:
RavenDB-23085 - Fixing IO Ring code
- Will properly consume the events for Posix IO Ring, avoid 100% CPU
- Will only wait for submitted events if the queue is full on Windows
- Making sure to advance the arena pointer on Windows

ayende committed:
RavenDB-23085
- Fixing how we want for eventfd by using raw read instead of poll + read.
- Ensuring that if we waited once, we'll reset the must_wait for the next run

ayende committed:
RavenDB-23085
 - Allowing to fsync a set of directories using IO Ring
- Refactoring to extract arena struct so we can have one for writes and one for fsync folders

ayende committed:
RavenDB-23085 - Wire the paths to sync from the managed side

ayende committed:
RavenDB-23085 - Handle the case when we flush directories, we'll not error if the directory does not exists (an index that was already removed)

ayende committed:
RavenDB-23085 Fixing compilation

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085
- Ensure that we are only adding the journals path to flush when we are using directory (persistent) env. Otherwise, it makes no sense

ayende committed:
RavenDB-23085 - Move the linked folders sync outside of the lock

ayende committed:
RavenDB-23759 Fix compilation on VS 17.13 on the v7.1 branch

ayende committed:
RavenDB-23756 Transaction Merger should use async commits by default

ayende committed:
RavenDB-23085 - Avoid checking if the latest journal file is linked the second time, to reduce the number of system calls that we use

ayende committed:
RavenDB-23085
- Moving shared journals to a dedicated "JournalsForIndexing" directory to avoid holding up the db transaction merger
- We'll now favor throughput for shared journals writes over latency

ayende committed:
RavenDB-23085 - Update the configuration of the index shared journals based of the database configuration, make sure to make a dummy commit every few journal commits so we'll "use" and cleanup the journals.

ayende commented on src/Raven.Server/Documents/Indexes/SharedIndexJournals.cs:
```diff
@@ -0,0 +1,112 @@
+    private bool _disposed;
+    private readonly PoolOfThreads.LongRunningWork _sharedJournalsThread;
+
+    private void WriteSharedJournals(object _)
```
@karmeli87 Need to sit on this and figure out what is the proper error handling strategy here is meant to be.

ayende committed:
RavenDB-23085 - Only generating a commit for shared journal instance if we changed the journal, to reduce amount of work needed

ayende committed:
RavenDB-23085 - Updating .NET SDK to 9.0.200

ayende committed:
RavenDB-23085 - Formatting the tryout code

ayende committed:
RavenDB-23085 - Fixing Symbol Reader for GC dump in Raven.Debug project.
Avoiding:
 System.MissingMethodException: Method not found: 'Void Microsoft.Diagnostics.Symbols.SymbolReader..ctor(System.IO.TextWriter, System.String)'.

ayende committed:
RavenDB-23085
- Fixing race condition where the shared index scope would be set only after indexing already started to run
- Making sure that we create the shared journal instance in indexes as part of restoring.
- Refactoring how we initialize the shared journals in indexes so we can also call it during snapshot restore

ayende committed:
RavenDB-23085 - Fixing potential race condition in Windows' IO Ring usage.
- When closing the ring in Windows - wait for all submitted entries to complete

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085 - We need to make sure that the order of operations when disposing a tx merger transaction is:
- First dispose the transaction (and run any OnDispose events)
- Notify all clients that the transaction has been disposed.

Otherwise, we may run into race conditions. This was ensured when we used sync commit, but not when using async. Moving to always async led to a problem, hence this fix

ayende committed:
RavenDB-23085 - Using .NET 9.0.103 - to match CI version

ayende committed:
RavenDB-23085
- Make sure that the windows IO Ring thread is properly named
- Changed error handling so if the ring had an error, we'll immediately respond with errors to callers, instead of having to wait for the thread.
- Generate PDBs on Windows for PAL

redknightlois commented on src/Raven.Pal/src/rvngetpalver.c:
```diff
@@ -1,3 +1,3 @@
 #include <sys/types.h>
 #include "rvn.h"
-EXPORT int32_t rvn_get_pal_ver() { return  62022 ; }
+EXPORT int32_t rvn_get_pal_ver() { return  70812 ; }
```
Shouldnt this be 71xxx?

ayende replied:
eventually, yes. Still making changes, and need to make sure we aren't consuming `7.1` in local nuget caches

arekpalinski replied:
https://issues.hibernatingrhinos.com/issue/RavenDB-24308/Set-Raven.PAL-to-8.0.0

redknightlois commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -318,7 +318,7 @@ protected dynamic TryConvert<T>(object value)
                     return f;
-                if (v is LazyStringValue lsv && double.TryParse(lsv, out var r))
+                if (v is LazyStringValue lsv && double.TryParse(lsv.AsReadOnlySpan(), out var r))
                     return r;
```
Shouldn't this be applied independently?

ayende replied:
Yes, but it won't compile otherwise :-)
Was merged seprately, IIRC

redknightlois commented on src/Sparrow/Json/IBlittableJsonTextWriter.cs:
```diff
@@ -24,6 +24,7 @@ public interface IBlittableJsonTextWriter
     void WritePropertyName(ReadOnlySpan<byte> prop);
 
+    void WritePropertyName(LazyStringValue prop);
     void WritePropertyName(string prop);
```
It makes sense for us to push this change to the v7.0 codebase to avoid divergencies at the Blittable interface. cc @arekpalinski

arekpalinski replied:
I looks that changes made here got reverted as there is nothing changed here

redknightlois commented on src/Voron/Impl/FileHeaders/FileHeader.cs:
```diff
@@ -1,14 +1,20 @@
+        public static int HashOffset = sizeof(FileHeader) - sizeof(ulong);
+        
+        public static int TransactionIdOffset = (int)Marshal.OffsetOf<FileHeader>(nameof(TransactionId));
 
```
These 2 can be marked static. By doing so we ensure that the JIT will inline them as constants.

ayende replied:
You mean `static readonly`, no?

redknightlois replied:
Yes. My bad, was thinking about readonly and wrote static.

ayende committed:
RavenDB-23085
- Working around compiler bug: https://developercommunity.visualstudio.com/t/MSVC-generate-non-thread-safe-code-when/10866952

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085 - Major refactoring of the ioring behavior on Windows.
- Will submit all work to the ring in one shot
- Will notify the submitter thread just once, instead of per operation
- Simplified error handling
- Making sure that we submit the IO operations to the OS in a sequential manner

ayende committed:
RavenDB-23085
- Refactoring IO ring usage on Linux as well
- Reduced contention, better cross thread signaling

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085
- Fixing submittion of pages to the ring, since we merge the writes, we need to pass the merged count, not the total
- Use the proper index for the start of the write when merging writes
- Make sure to pass the submittion properly during folder sync
- Simplified multi folder sync and ensures that we pass the proper list to queue_work()

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085
- Fixing batches that use > IOV_MAX number of pages
- Added failing test

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085 - Fixing the build

ayende committed:
RavenDB-23085
- Removing txMergerShutdown cancellation token - no longer needed
- Making sure that the scope for shared journal match the scope of the shared journals thread
- The shared journal scope use its own cancellation token, to avoid external factors cancelling it out of turn

ayende committed:
RavenDB-23085 - Removing old unused binaries

ayende committed:
RavenDB-23085 - Adding PAL reference to the EmbeddedTests project (required for zstd reference)

ayende committed:
RavenDB-23085
- Fixing test (We no longer shared index journals with database journal)
- Ensuring that we use a _new_ journal for all indexes by forcing indexing

ayende committed:
RavenDB-23085 - Wiring the shared journals when reading existing indexes on the disk

ayende committed:
RavenDB-23085
- Removed the default created 000000.journal file if we restored on an empty directory. The single tx in it is using the wrong journal id.

ayende committed:
RavenDB-23085 - Fixing polarity of test

ayende committed:
RavenDB-23085 - Properly detect missing journal files for Voron's incremental backups

karmeli87 committed:
RavenDB-23085 fix tests using the LatestJournalNumber

karmeli87 committed:
RavenDB-23085 fix encryption of the linked journals

karmeli87 committed:
RavenDB-23085
- adjust the calling conventions to C-style by adding [UnmanagedCallersOnly(CallConvs = [typeof(CallConvCdecl)])]
- `bool` isn't marshable, so use `int` instead

karmeli87 committed:
RavenDB-23085 dispose the journal only after reader has completed to avoid corruption

karmeli87 committed:
RavenDB-23085 jounral recovery should ensure there are no gaps in the journals files

karmeli87 committed:
RavenDB-23085
- add the env jounral id to the linked entry
- upon recover read the header of the env and compare to the one in the linked entry
This will ensure that the same env being used and avoid inconsistency when index is reset or manually deleted

karmeli87 committed:
RavenDB-23085
- work with the env options directly
- handle recovery in the case when the first journal is missing

karmeli87 committed:
RavenDB-23085 ensure we can disable shard journals

karmeli87 committed:
RavenDB-23085
- unify reading the header upon shared journal recovery
- tests that delete the index journal to simulate failure, must also delete the shared journal folder to avoid recovery.

karmeli87 committed:
RavenDB-23085 fix tests and add debug info

ayende committed:
RavenDB-23085 - Fixing order of statements - we accidentally zero-init the mutexes, which only worked on Linux

ayende committed:
RavenDB-23085 - Updating binaries

karmeli87 committed:
RavenDB-23085 simplify test logic for not running in memory

karmeli87 committed:
RavenDB-23085
- remove zstd from PAL
- no need for the DynamicNativeLibraryResolver
- clean up an unused ZstdSharp.Port package

karmeli87 committed:
RavenDB-23085 Updating binaries

karmeli87 committed:
RavenDB-23085 adjust GetDocumentStore logic

karmeli87 committed:
RavenDB-23085 fix build process

karmeli87 committed:
RavenDB-23085 fix libzstd

karmeli87 committed:
RavenDB-23085 fix complication

karmeli87 committed:
RavenDB-23085 handle failures of during shared journal commits

arekpalinski commented on src/Raven.Server/Documents/Indexes/SharedIndexJournals.cs:
```diff
@@ -0,0 +1,134 @@
+                        using (var txw = _env.WriteTransaction())
+                        {
+                            txw.Commit();
+                        }
```
What is the purpose of this empty, write transaction?

arekpalinski commented on src/Raven.Server/Documents/Indexes/SharedIndexJournals.cs:
```diff
@@ -0,0 +1,134 @@
+{
+    private readonly DocumentDatabase _documentDatabase;
+
+    public SharedIndexJournals(DocumentDatabase documentDatabase)
```
What is `SharedIndexJournals` about? I see it creates a separate `StorageEnvironment` per each database.

arekpalinski commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -348,7 +348,7 @@ private async Task<bool> ProxyToLeaderIfNeeded(JsonOperationContext context, Dat
         }
 
-        private void RecreateIndexes(string databaseName, DatabaseRecord databaseRecord)
+        private async Task RecreateIndexes(string databaseName, DatabaseRecord databaseRecord)
```
`RecreateIndexesAsync`

arekpalinski replied:
done

arekpalinski commented on src/Voron/Impl/Journal/JournalFile.cs:
```diff
@@ -6,62 +6,82 @@
+                    continue;
+                count++;
+                first = Math.Min(_transactionHeaders[i].TransactionId, first);
+                last = _transactionHeaders[i].TransactionId;
```
We're looking for the lowest tx id because they might come from different environments, right? So we have `first = Math.Min()`. Why we don't do `Math.Max()` for `last`?

ayende replied:
We can change `first` set to be:  
```
if (first == long.MaxValue) 
                first = _transactionHeaders[i].TransactionId;
```

It would do the same.

Basically, once we set it to a value, we want to keep it like that.

arekpalinski replied:
done

arekpalinski commented on src/Voron/Impl/Journal/JournalFile.cs:
```diff
@@ -150,19 +153,86 @@ public void Write(LowLevelTransaction tx, CompressedPagesResult pages)
+        /// - There *are* existing transactions in the journal *and* that journal is a hard link
+        ///
+        /// The issue is that we may have a snapshot / manual file move that would result in breaking
+        /// of the hard link between shared journals. Consider the case of an index & database that have
```
1. Will it still be possible to get raw data of an index (and shared journals) and inspect it?
2. Will it be possible to get just a database data (and shared journals) and load it without indexes? Won't it fail on the recovery due to encountering transactions coming from indexes environments?

arekpalinski commented on src/Voron/Impl/Journal/JournalInfo.cs:
```diff
@@ -14,8 +14,13 @@ public unsafe struct JournalInfo
+        /// </summary>
         [FieldOffset(0)]
-        public long CurrentJournal;
+        public long Reserved1;
```
Will there be an option to _not_ use the shared journals in 7.1? So it would work the same as before 7.1?

ayende replied:
You can _not_ use it, yes.
However, we are no longer looking at this field any more, even in that scenario

arekpalinski commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -559,107 +461,64 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, ref Pag
+                if (txIdDiff != 1)
                 {
-                    if (CanIgnoreDataIntegrityErrorBecauseTxWasSynced(current, options))
+                    if (CanIgnoreDataIntegrityErrorBecauseTxWasSynced(current->TransactionId -1 , options))
```
Why `-1`?

ayende replied:
The _current_ transaction is valid, but the _previous_ one may not have been.
That is what we are checking, whatever there are gaps, and since we need to know if we previously synced, we check for the tail end of the range, which is `-1` from the current known good end of gap

arekpalinski commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -671,19 +530,191 @@ void AssertValidLastPageNumber(TransactionHeader* transactionHeader)
+                    }
+                    // Here we are dealing with a valid transaction (in terms of tx id)
+                    // that has zeroed DatabaseId, probably a legacy transaction for the 
+                    // current database (non-shared journal mode), allowing it
```
How the legacy dbs are going to work? After recovering they will start using shared journal mode automatically, right?

Do we ensure that the legacy journals will be removed during immediately after the recovery? So the next attempt to do the recovery will not attempt to operate in legacy mode

ayende replied:
Legacy dbs just reads their journals as normal.
On new commit, they will create a new linked journal, the old one is still in use.

arekpalinski commented on src/Voron/Impl/Journal/TransactionHeader.cs:
```diff
@@ -74,8 +74,12 @@ public unsafe struct TransactionHeader
+        /// transaction belongs to. 
+        /// </summary>
+        [FieldOffset(136)]
+        public Guid JournalId;
```
Maybe lets also put  `public fixed byte Reserved[16];` to the unused bytes

arekpalinski commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -1963,71 +2365,6 @@ private void EncryptTransaction(byte* fullTxBuffer)
                 throw new InvalidOperationException("Failed to call crypto_aead_xchacha20poly1305_ietf_encrypt, rc = " + rc);
         }
 
-        private sealed class CompressionAccelerationStats
```
Is it something no longer valuable for shared journals only or we should consider removing it in 7.0 as well?

ayende replied:
That wasn't really valuable, and we should scrap it. Having said that, not much pointing touching this code.

ppekrol commented on libs/liburing/readme.txt:
```diff
@@ -0,0 +1,8 @@
+The files were taken from:
+
+https://launchpad.net/ubuntu/+source/liburing/2.8-1
```
do we want to update this to https://launchpad.net/ubuntu/+source/liburing/2.9-1 ?

ayende replied:
Looks like all the changes there are not relevant for us

ppekrol commented on Directory.Packages.props:
```diff
@@ -94,7 +94,7 @@
     <PackageVersion Include="RabbitMQ.Client" Version="6.8.1" />
     <PackageVersion Include="Raven.CodeAnalysis" Version="1.0.11" />
-    <PackageVersion Include="Raven.Pal" Version="6.2.22" />
+    <PackageVersion Include="Raven.Pal" Version="7.0.835" />
```
7.1...

ayende replied:
That is for just before the release - I want to have a nice "7.1.0" for the actual release.
Right now, we churn through those _very_ quickly, and it may "poison" the nuget cache

arekpalinski replied:
https://issues.hibernatingrhinos.com/issue/RavenDB-24308/Set-Raven.PAL-to-8.0.0

ppekrol commented on src/Raven.Pal/Raven.Pal.vcxproj:
```diff
@@ -46,7 +49,7 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="Configuration">
     <ConfigurationType>Application</ConfigurationType>
-    <UseDebugLibraries>false</UseDebugLibraries>
+    <UseDebugLibraries>true</UseDebugLibraries>
```
assuming that this is not a problem, because we are building via script anyway?

ayende replied:
Yes, that is not actually how we build, yes.

ppekrol commented on src/Raven.Pal/Raven.Pal.vcxproj:
```diff
@@ -19,14 +20,16 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Label="Configuration">
-    <ConfigurationType>DynamicLibrary</ConfigurationType>
+    <ConfigurationType>Application</ConfigurationType>
```
assuming that this is not a problem, because we are building via script anyway?

ayende replied:
Correct

ppekrol commented on src/Raven.Pal/build.ps1:
```diff
@@ -51,39 +49,58 @@ $posix_files = `
+    zig cc -Wall -O3 -g -shared -fPIC -Iinc -target x86-windows -o runtimes/win-x86/native/librvnpal.dll $shared $win_files 
+}
+else {
+    cmd /c """$vcbin\vcvars32.bat"" & ""$clbin\x86\cl"" /Fdruntimes/win-x86/native/librvnpal.pdb /Zi -Feruntimes/win-x86/native/librvnpal.dll -I inc /O2 /analyze /sdl /LD $shared $win_files /link"
```
why not zig? I'm curious

ayende replied:
I would _love_ to use Zig here.
But I run into a problem with it requiring a certain version of Windows, and I got stuck with that.
see: https://www.reddit.com/r/Zig/comments/1h73sft/zig_cc_avoiding_linking_to_the_c_library/

ppekrol commented on src/Raven.Pal/pal.ver:
```diff
@@ -1 +1 @@
-62022
+70835
```
71xxx

arekpalinski replied:
https://issues.hibernatingrhinos.com/issue/RavenDB-24308/Set-Raven.PAL-to-8.0.0

ppekrol commented on src/Raven.Pal/src/rvngetpalver.c:
```diff
@@ -1,3 +1,3 @@
 #include <sys/types.h>
 #include "rvn.h"
-EXPORT int32_t rvn_get_pal_ver() { return  62022 ; }
+EXPORT int32_t rvn_get_pal_ver() { return  70835 ; }
```
71xxx

arekpalinski replied:
This is set automatically bu the build script - https://issues.hibernatingrhinos.com/issue/RavenDB-24308/Set-Raven.PAL-to-8.0.0

ppekrol commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -163,8 +161,11 @@ public virtual bool RunInMemory
+        public virtual PathSetting StoragePath => field ??= _root.ResourceType == ResourceType.Server ? null : _root.Core.DataDirectory.Combine("Indexes");
 
+        [ReadOnlyPath]
+        public virtual PathSetting SharedJournalsPath => field ??= _root.ResourceType == ResourceType.Server ? null : _root.Core.DataDirectory.Combine("JournalsForIndexing");
```
we do not want to keep them inside `Indexes` ?

ayende replied:
If you put it inside indexes, you have to be concerned about an index called `SharedJournals`

arekpalinski replied:
@ayende what if we could go with `@SharedJournals` put inside `Indexes` directory? `@` character isn't allowed in an index name so we won't need to worry about index named "SharedJournals". 

This way we'd have all index related data in the same directory. It makes operations on raw data files easier, e.g. copying just docs data means excluding only `Indexes` directory, forcing to recreate all indexes would also be just removal of `Indexes` directory.

`@SharedJournals` would be first / on top in `Indexes` directory, so easier to find it than just `SharedJournals`

ayende replied:
I'm fine with doing that, yes.

arekpalinski replied:
done

ppekrol commented on src/Raven.Server/Config/Categories/StorageConfiguration.cs:
```diff
@@ -70,6 +71,12 @@ public StorageConfiguration()
+        [DefaultValue(2048)]
+        [SizeUnit(SizeUnit.Megabytes)]
+        [ConfigurationEntry("Storage.MaxJournalFileSizeInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
+        public Size MaxJournalFileSize { get; set; }
```
should be smaller for x86?

ayende replied:
I... don't think so.
For x86, that is just the file size.
During startup, we run through all those transactions on a per file basis, but I don't think we keep a lot of that in memory.
It is not a _bad_ idea, we don't care that much about x86 performance.

ppekrol commented on src/Raven.Server/Config/Categories/StorageConfiguration.cs:
```diff
@@ -147,15 +159,37 @@ public StorageConfiguration()
+        [DefaultValue(false)]
+        [ConfigurationEntry("Storage.LowPriorityFlushAndSync", ConfigurationEntryScope.ServerWideOnly)]
+        public bool LowPriorityFlushAndSync { get; set; }
+        
```
spacing

arekpalinski replied:
done

ppekrol commented on src/Raven.Server/Config/Categories/StorageConfiguration.cs:
```diff
@@ -147,15 +159,37 @@ public StorageConfiguration()
+        
+        
+        [Description("EXPERT: The queue size to use for I/O ring operations, using -1 will disable I/O ring entirely")]
+        [DefaultValue(1024)]
```
do we need less for x86?

ayende replied:
No, it isn't relevant on x86 at all

ppekrol commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -605,6 +606,24 @@ public static Index Open(string path, DocumentDatabase documentDatabase, bool ge
+                return;
+            }
+
+            var logger = documentDatabase.Loggers.GetLogger(typeof(Index));
```
this is incorrect, we need to implement 'GetLoggerForIndex' (check impl in raven log manager extensions) for DatabaseLoggersContext

ppekrol commented on src/Raven.Studio/typescript/models/database/index/configurationItem.ts:
```diff
@@ -49,7 +49,9 @@ class configurationItem {
-        "Indexing.Corax.VectorSearch.OrderByScoreAutomatically"
+        "Indexing.Corax.VectorSearch.OrderByScoreAutomatically",
+        "Storage.DisableSparseRegions",
+        "Storage.JournalsCompressionAcceleration"
```
anything else needs to be added here?

ppekrol commented on src/Raven.Studio/typescript/viewmodels/manage/storageReport.ts:
```diff
@@ -233,7 +233,7 @@ class storageReport extends viewModelBase {
             );
             
-            item.recyclableJournal = temp.Type === "RecyclableJournal";
+            item.recyclableJournal = false;
```
I think we need to make adjustments to storage report now?

ayende replied:
Yes

ppekrol commented on src/Sparrow.Server/Platform/MemoryLockUsage.cs:
```diff
@@ -0,0 +1,82 @@
+        private static long GetNearestFileSize(long neededSize)
+        {
+            if (neededSize < IncreaseByPowerOf2Threshold)
+                return Bits.PowerOf2(neededSize);
```
should this be Bits.NextAllocationSize ?

ayende replied:
Probably

arekpalinski replied:
It's okay here. We have `if (neededSize < IncreaseByPowerOf2Threshold)` so we grow by power of 2 only below 512MB

ppekrol commented on src/Sparrow/Utils/ZstdLib.cs:
```diff
@@ -18,6 +18,7 @@ internal static unsafe class ZstdLib
 
         static ZstdLib()
         {
+            DynamicNativeLibraryResolver.Register(typeof(ZstdLib).Assembly, LIBZSTD);
```
todo @ppekrol

ayende replied:
?

ppekrol replied:
This is for me, ignore please
